File: Android/shared/src/main/kotlin/shared/dto/objects/Position.kt
@file:UseSerializers(UUIDSerializer::class)

package shared.dto.objects

import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import shared.dto.optionSet.PermissionSet
import shared.dto.protocols.Parametable
import shared.dto.protocols.Responsable
import shared.serializers.UUIDSerializer
import java.util.*

object Position {
    data object Parameters {
        @Serializable
        data class Create(
            val title: String,
            val permissions: PermissionSet,
            val salary: Salary.Parameters.Rules.Create,
        ) : Parametable()

        @Serializable
        data class Patch(
            val title: String?,
            val permissions: PermissionSet?,
            val salary: Salary.Parameters.Rules.Create,
        ) : Parametable()
    }

    data object Responses {
        @Serializable
        data class Full(
            val id: UUID,
            val title: String,
            val permissions: PermissionSet,
            val salary: Salary.Responses.Rules.Full,
        ) : Responsable

        @Serializable
        data class Partial(
            val id: UUID,
            val title: String,
        ) : Responsable
    }
}

File: Node/src/dto/objects/Position.ts
import { PermissionSet } from "../optionSet/PermissionSet"
import { Salary } from "./Salary"
import { UUID } from "../tsPrimitives/UUID"
import { ValidateNested } from "class-validator"

export namespace Position {

    export namespace Parameters {
        
        export class Create {
            title: string

            @ValidateNested()
            permissions: PermissionSet

            @ValidateNested()
            salary: Salary.Parameters.Rules.Create

            constructor (
                title: string,
                permissions: PermissionSet,
                salary: Salary.Parameters.Rules.Create
            ) {
                this.title = title
                this.permissions = permissions
                this.salary = salary
            }
        }

        export class Patch {
            title?: string

            @ValidateNested()
            permissions?: PermissionSet

            @ValidateNested()
            salary?: Salary.Parameters.Rules.Create

            constructor (
                title?: string,
                permissions?: PermissionSet,
                salary?: Salary.Parameters.Rules.Create
            ) {
                this.title = title
                this.permissions = permissions
                this.salary = salary
            }
        }
    }

    export namespace Responses {
        
        export class Full {
            title: string

            @ValidateNested()
            id: UUID

            @ValidateNested()
            permissions: PermissionSet

            @ValidateNested()
            salary: Salary.Responses.Rules.Full

            constructor (
                id: UUID,
                title: string,
                permissions: PermissionSet,
                salary: Salary.Responses.Rules.Full
            ) {
                this.id = id
                this.title = title
                this.permissions = permissions
                this.salary = salary
            }
        }

        export class Partial {
            title: string

            @ValidateNested()
            id: UUID

            constructor (
                id: UUID,
                title: string
            ) {
                this.id = id
                this.title = title
            }
        }
    }
}